### ISSUE TITLE
Segfault while trying to write restart at ne1024 on pm-gpu/pm-cpu

### ISSUE THREAD

**Issue by ndkeen** [E3SM-Project/scream#2042]
With Nov16th scream repo, I see a generic segfault while trying to write restarts at ne1024 on pm-gpu.
This is after adding the MPICH env var to use barrier before each bcast (already in e3sm master).
And I've verified that ne30/ne120/ne256 cases on both pm-gpu/pm-cpu work, but had
not yet tried ne1024 restarts.

```
1204: Program received signal SIGSEGV: Segmentation fault - invalid memory reference.
1204:
1204: Backtrace for this error:
1204: #0  0x148b025e7d6f in ???
1204: #1  0x148b064aecc2 in ???
srun: error: nid002624: task 1204: Segmentation fault
srun: launch/slurm: _step_signal: Terminating StepId=3693094.0
   0: slurmstepd: error: *** STEP 3693094.0 ON nid001076 CANCELLED AT 2022-11-18T13:05:50 ***

/pscratch/sd/n/ndk/e3sm_scratch/pm-gpu/se42-nov16/f1024.F2010-SCREAMv1.ne1024pg2_ne1024pg2.se42-nov16.gnugpu.1d.n0384a4xX18888.wo216n1.wr.s64
```

it looks like it fails trying to write `*scream.r.0001-01-02-00000.nc` file (which has zero size).

I have 2 cases with same error (both 384 pm-gpu nodes).  Both cases completed all of the steps for 1 day and wrote other output/restarts to that point. I've also ran several 1-day cases without restarts.

@AaronDonahue reminded me that this could be the very same issue that happened on summit at ne1024 and was resolved with a patch to scorpio.

```
cd /path/to/your/scream/source/directory
cd externals/scorpio
git fetch origin
git merge --no-edit origin/dqwu/scorpio_1.3.2_patch
```

I have a job in the Q with this fix (which is not yet in E3SM master).
`/pscratch/sd/n/ndk/e3sm_scratch/pm-gpu/se43-nov18/f1024.F2010-SCREAMv1.ne1024pg2_ne1024pg2.se43-nov18.gnugpu.1d.n0384a4xX.no216n8.s32.wr`

**Comment by ndkeen** [comment]
Trying the scorpio 1.3.2 patch (with Nov 18th scream repo), I still see the same issue.
It fails with segfault and based on the output directory, it looks like it happened just after opening
the `*.scream.r.0001-01-02-00000.nc` file.

`/pscratch/sd/n/ndk/e3sm_scratch/pm-gpu/se43-nov18/f1024.F2010-SCREAMv1.ne1024pg2_ne1024pg2.se43-nov18.gnugpu.1d.n0384a4xX.prod.ybo.wr.s32`

**Comment by jayeshkrishna** [comment]
Did you try out the env vars suggested by NERSC (https://github.com/E3SM-Project/E3SM/pull/5275) ?

**Comment by ndkeen** [comment]
Not yet.  I can try another job with those env vars as well.  Note they were actually suggested by HPE. NERSC was just messenger.

**Comment by ndkeen** [comment]
I also just tried on pm-cpu and see what looks like the same problem.  It stops after opening the scream restart file.

```
2122:
2122: Program received signal SIGBUS: Access to an undefined portion of a memory object.
2122:
2122: Backtrace for this error:
2122: #0  0x152ef623dd6f in ???
2122: #1  0x16b50b6 in ???
2122: #2  0x152efa9ada55 in gomp_thread_start
```

**Comment by ndkeen** [comment]
With Nov 18th repo (using the `dqwu/scorpio_1.3.2_patch` branch), I did get one case to complete the restart write, but I had to use 512 nodes. So it has failed with a few attempts at 384 nodes, a few times with 400 nodes, once with 448 nodes, and also failed once or twice with 512 nodes -- but did complete writing restarts twice with 512 nodes. I am doubtful the issue here is simply not having enough memory -- it seems more like a race condition.

I also tried ne1024 cases with Nov 23rd repo which failed (384 and 400-node attempts).

And then I tried with Nov 24th repo _and_ using the latest PIO -- ie in `externals/scorpio`, doing `git pull origin master` to see hash `c98c351d`.  With this, I see one fail with 400 nodes, 2 fails with 512 nodes, and 1 completed restart write with 512 nodes.

**Comment by ndkeen** [comment]
After PM maintenance, and https://github.com/E3SM-Project/E3SM/pull/5373, I tried this case again with Dec 26th repo.  It ran for 1 day on 348 pm-gpu nodes without restart. When I tried adding a restart, it fails in the same way. And that's with my changes to NOT use MPICH on GPU. It didn't seem like this error was related to https://github.com/E3SM-Project/scream/pull/2102, but I tried it anyway.

**Comment by ndkeen** [comment]
I tried again with Jan4th repo -- on pm-gpu used 384 nodes, same segfault right after opening the `scream.r` file.  But then I tried a 512 node job and it completed 1 day with restart. Note this also happened before, where most of my 512 jobs were completing, but I did have some that failed with same segfault.

It seems odd that writing a restart requires 33% more memory.

Note that pm-gpu has 40GB of memory per A100 (160GB total per node) and 256GB of memory for CPU.
